From f460fda131b227ffc5734000e273b9eef475cb63 Mon Sep 17 00:00:00 2001
From: Hans de Goede <johannes.goede@oss.qualcomm.com>
Date: Thu, 4 Dec 2025 21:44:05 +0100
Subject: [PATCH 415/419] Revert "arm64: Fix EFI loader kernel image
 allocation"

This fully reverts commit 3cfd4a1def0d13104a6ae6a89672bc48102498cd.

Note the modifed code is not used on x86 (#ifdef !x86).

This is part of a series of reverts to move non x86 vmlinuz loading back
to using EFI boot services load_image + start_image calls.

The current DYI approach breaks booting kernel EFI binaries which have
the stubble stub for automatic DTB loading.

Switching back to EFI boot services load_image + start_image fixes this
and also brings Fedora's grub inline with how upstream grub is loading
and booting kernels on arm64 platforms.

Signed-off-by: Hans de Goede <johannes.goede@oss.qualcomm.com>
---
 grub-core/loader/efi/linux.c | 93 +++++++++++-------------------------
 include/grub/efi/efi.h       |  3 +-
 2 files changed, 28 insertions(+), 68 deletions(-)

diff --git a/grub-core/loader/efi/linux.c b/grub-core/loader/efi/linux.c
index ac246bb88..720d5b02b 100644
--- a/grub-core/loader/efi/linux.c
+++ b/grub-core/loader/efi/linux.c
@@ -42,8 +42,6 @@ GRUB_MOD_LICENSE ("GPLv3+");
 static grub_dl_t my_mod;
 static int loaded;
 
-static void *kernel_alloc_addr;
-static grub_uint32_t kernel_alloc_pages;
 static void *kernel_addr;
 static grub_uint64_t kernel_size;
 static grub_uint32_t handover_offset;
@@ -561,8 +559,9 @@ grub_linux_unload (void)
 			 GRUB_EFI_BYTES_TO_PAGES (initrd_end - initrd_start));
   initrd_start = initrd_end = 0;
   grub_free (linux_args);
-  if (kernel_alloc_addr)
-    grub_efi_free_pages ((grub_addr_t) kernel_alloc_addr, kernel_alloc_pages);
+  if (kernel_addr)
+    grub_efi_free_pages ((grub_addr_t) kernel_addr,
+			 GRUB_EFI_BYTES_TO_PAGES (kernel_size));
 #if !defined(__i386__) && !defined(__x86_64__)
   grub_fdt_unload ();
 #endif
@@ -753,37 +752,12 @@ grub_cmd_initrd (grub_command_t cmd __attribute__ ((unused)),
   return grub_errno;
 }
 
-#if !defined(__i386__) && !defined(__x86_64__)
-static grub_err_t
-parse_pe_header (void *kernel, grub_uint64_t *total_size,
-		 grub_uint32_t *entry_offset,
-		 grub_uint32_t *alignment)
-{
-  struct linux_arch_kernel_header *lh = kernel;
-  struct grub_armxx_linux_pe_header *pe;
-
-  pe = (void *)((unsigned long)kernel + lh->hdr_offset);
-
-  if (pe->opt.magic != GRUB_PE32_PEXX_MAGIC)
-    return grub_error(GRUB_ERR_BAD_OS, "Invalid PE optional header magic");
-
-  *total_size   = pe->opt.image_size;
-  *entry_offset = pe->opt.entry_addr;
-  *alignment    = pe->opt.section_alignment;
-
-  return GRUB_ERR_NONE;
-}
-#endif
-
 static grub_err_t
 grub_cmd_linux (grub_command_t cmd __attribute__ ((unused)),
 		int argc, char *argv[])
 {
   grub_file_t file = 0;
   struct linux_arch_kernel_header lh;
-  grub_off_t filelen;
-  grub_uint32_t align;
-  void *kernel = NULL;
   grub_err_t err;
 
   grub_dl_ref (my_mod);
@@ -814,6 +788,8 @@ grub_cmd_linux (grub_command_t cmd __attribute__ ((unused)),
   if (!file)
     goto fail;
 
+  kernel_size = grub_file_size (file);
+
   if (grub_arch_efi_linux_load_image_header (file, &lh) != GRUB_ERR_NONE)
 #if !defined(__i386__) && !defined(__x86_64__)
     goto fail;
@@ -834,46 +810,33 @@ fallback:
     }
 #endif
 
-  filelen = grub_file_size (file);
-  kernel = grub_malloc(filelen);
-  if (!kernel)
-    {
-      grub_error (GRUB_ERR_OUT_OF_MEMORY, N_("cannot allocate kernel load buffer"));
-      goto fail;
-    }
-
-  if (grub_file_read (file, kernel, filelen) < (grub_ssize_t)filelen)
-    {
-      grub_error (GRUB_ERR_FILE_READ_ERROR, N_("Can't read kernel %s"),
-		  argv[0]);
-      goto fail;
-    }
-
-#if !defined(__i386__) && !defined(__x86_64__)
-  if (parse_pe_header (kernel, &kernel_size, &handover_offset, &align) != GRUB_ERR_NONE)
-    goto fail;
-  grub_dprintf ("linux", "kernel mem size     : %lld\n", (long long) kernel_size);
-  grub_dprintf ("linux", "kernel entry offset : %d\n", handover_offset);
-  grub_dprintf ("linux", "kernel alignment    : 0x%x\n", align);
-
   grub_loader_unset();
 
-  kernel_alloc_pages = GRUB_EFI_BYTES_TO_PAGES (kernel_size + align - 1);
-  kernel_alloc_addr = grub_efi_allocate_any_pages (kernel_alloc_pages);
-  grub_dprintf ("linux", "kernel numpages: %d\n", kernel_alloc_pages);
-  if (!kernel_alloc_addr)
+  grub_dprintf ("linux", "kernel file size: %lld\n", (long long) kernel_size);
+  kernel_addr = grub_efi_allocate_any_pages (GRUB_EFI_BYTES_TO_PAGES (kernel_size));
+  grub_dprintf ("linux", "kernel numpages: %lld\n",
+		(long long) GRUB_EFI_BYTES_TO_PAGES (kernel_size));
+  if (!kernel_addr)
     {
       grub_error (GRUB_ERR_OUT_OF_MEMORY, N_("out of memory"));
       goto fail;
     }
-  kernel_addr = (void *)ALIGN_UP((grub_uint64_t)kernel_alloc_addr, align);
+
+  grub_file_seek (file, 0);
+  if (grub_file_read (file, kernel_addr, kernel_size)
+      < (grub_int64_t) kernel_size)
+    {
+      if (!grub_errno)
+	grub_error (GRUB_ERR_BAD_OS, N_("premature end of file %s"), argv[0]);
+      goto fail;
+    }
 
   grub_dprintf ("linux", "kernel @ %p\n", kernel_addr);
-  grub_memcpy (kernel_addr, kernel, grub_min(filelen, kernel_size));
-  if (kernel_size > filelen)
-    grub_memset ((char *)kernel_addr + filelen, 0, kernel_size - filelen);
-  grub_free(kernel);
-  kernel = NULL;
+
+#if !defined(__i386__) && !defined(__x86_64__)
+  struct grub_armxx_linux_pe_header *pe;
+  pe = (void *)((unsigned long)kernel_addr + lh.hdr_offset);
+  handover_offset = pe->opt.entry_addr;
 #endif
 
   cmdline_size = grub_loader_cmdline_size (argc, argv) + sizeof (LINUX_IMAGE);
@@ -898,9 +861,6 @@ fallback:
     }
 
 fail:
-  if (kernel)
-    grub_free (kernel);
-
   if (file)
     grub_file_close (file);
 
@@ -913,8 +873,9 @@ fail:
   if (linux_args && !loaded)
     grub_free (linux_args);
 
-  if (kernel_alloc_addr && !loaded)
-    grub_efi_free_pages ((grub_addr_t) kernel_alloc_addr, kernel_alloc_pages);
+  if (kernel_addr && !loaded)
+    grub_efi_free_pages ((grub_addr_t) kernel_addr,
+			 GRUB_EFI_BYTES_TO_PAGES (kernel_size));
 
   return grub_errno;
 }
diff --git a/include/grub/efi/efi.h b/include/grub/efi/efi.h
index 7eed1bd79..4bd196500 100644
--- a/include/grub/efi/efi.h
+++ b/include/grub/efi/efi.h
@@ -20,6 +20,7 @@
 #ifndef GRUB_EFI_EFI_HEADER
 #define GRUB_EFI_EFI_HEADER	1
 
+#include <grub/efi/pe32.h>
 #include <grub/types.h>
 #include <grub/dl.h>
 #include <grub/efi/api.h>
@@ -51,12 +52,10 @@ struct grub_arm64_linux_pe_header
 };
 
 #if defined(__arm__)
-# define GRUB_PE32_PEXX_MAGIC GRUB_PE32_PE32_MAGIC
 # define grub_armxx_linux_pe_header grub_arm_linux_pe_header
 #endif
 
 #if defined(__aarch64__)
-# define GRUB_PE32_PEXX_MAGIC GRUB_PE32_PE64_MAGIC
 # define grub_armxx_linux_pe_header grub_arm64_linux_pe_header
 #endif
 
-- 
2.52.0

